src/backend/pipeline/fss.c

```diff
@@ -0,0 +1,455 @@
+/*-------------------------------------------------------------------------
+ *
+ * fss.c
+ *	  Filtered Space Saving implementation
+ *
+ *	  http://www.l2f.inesc-id.pt/~fmmb/wiki/uploads/Work/misnis.ref0a.pdf
+ *
+ * Copyright (c) 2013-2015, PipelineDB
+ *
+ * src/backend/pipeline/fss.c
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+
+#include <math.h>
+#include "pipeline/fss.h"
+#include "pipeline/miscutils.h"
+#include "utils/elog.h"
+#include "utils/palloc.h"
+
+#define DEFAULT_SS_M_FACTOR 6.0
+#define DEFAULT_M_FACTOR (DEFAULT_SS_M_FACTOR / 2)
+#define DEFAULT_H_FACTOR (DEFAULT_M_FACTOR * 6)
+
+#define MURMUR_SEED 0x02cd1b4c451c1fb8L
+
+FSS *
+FSSCreateWithMAndH(uint64_t k, TypeCacheEntry *typ, uint64_t m, uint64_t h)
+{
+	Size sz = sizeof(FSS) + (sizeof(Counter) * h) + (sizeof(MonitoredElement) * m);
+	char *pos;
+	FSS *fss;
+
+	if (k > m)
+		elog(ERROR, "maximum value of k exceeded");
+
+	/* TODO(usmanm): Check bounds for k, m, h */
+
+	/* TODO(usmanm): Add support for ref types */
+	if (!typ->typbyval)
+		elog(ERROR, "fss doesn't support types by reference");
```

Comment:
Yea, trying to figure out how to make this work. There's not way to tell the parser to only allow types by reference, so need some hackery.